File: Java/POO1/fraccion/Fraccion.java
package fraccion;

import javax.swing.JOptionPane;

/**
 * 5. Crea una clase Fraccion (en Java yPython) de forma que podamos hacer las
 * siguientes operaciones:
 * 
 * Contruir un objeto Fraccion pasándole al constructor el numerador y el
 * denominador. Obtener la fracción. Obtener y modificar numerador y
 * denominador. No se puede dividir por cero. Obtener resultado de la fracción
 * (número real). Multiplicar la fracción por un número. Multiplicar, sumar y
 * restar fracciones. Simplificar la fracción.
 * 
 * @author Juan Manuel Gonzalez Profumo
 * @version 1.0
 */
public class Fraccion {
	private int numerador;
	private int denominador;

	public Fraccion(int numerador, int denominador) {
		this.numerador = numerador;

		if (denominador == 0) {
			System.out.println("El denominador no puede ser 0");
		}

		else {
			this.denominador = denominador;
		}
	}

	public int getNumerador() {
		return numerador;
	}

	public void setNumerador(int numerador) {
		this.numerador = numerador;
	}

	public int getDenominador() {
		return denominador;
	}

	public void setDenominador(int denominador) {
		this.denominador = denominador;
	}

	Fraccion modificar(int n, int d) {
		return new Fraccion(n, d);
	}

	double calcularResultado() {
		return (double) numerador / denominador;
	}

	Fraccion multiplicarPorNumero(int num) {
		return new Fraccion(this.numerador * num, this.denominador);
	}

	Fraccion multiplicarFracciones(Fraccion fraccion2) {
		return new Fraccion(this.numerador * fraccion2.getNumerador(), this.denominador * fraccion2.getDenominador());
	}

	Fraccion sumarFracciones(Fraccion fraccion2) {
		return new Fraccion((this.numerador * fraccion2.getDenominador() + this.denominador * fraccion2.getNumerador()),
		    this.denominador * fraccion2.getDenominador());
	}

	Fraccion restarFracciones(Fraccion fraccion2) {
		return new Fraccion((this.numerador * fraccion2.getDenominador() - this.denominador * fraccion2.getNumerador()),
		    this.denominador * fraccion2.getDenominador());
	}

	public int hallarMcd() {
		int dividendo = this.numerador;
		int divisor = this.denominador;
		while (dividendo % divisor != 0) {
			int resto = dividendo % divisor;
			dividendo = divisor;
			divisor = resto;
		}

		return divisor;
	}

	public Fraccion simplificarFraccion() {
		int mcd = hallarMcd();
		this.numerador /= mcd;
		this.denominador /= mcd;
		return new Fraccion(numerador, denominador);
	}

	@Override
	public String toString() {
		return getNumerador() + "/" + getDenominador();
	}

}

File: Python/POO1/fraccion/Fraccion.py
class Fraccion():

    def __init__(self, num, den):
        self.__numerador = num
        self.__denominador = den

    def getNumerador (self):
        return self.__numerador

    def setNumerador (self, num):
        self.__numerador = num

    def getDenominador (self):
        return self.__denominador

    def setDenominador (self, den):
        self.__denominador = den

    def modificar (self, n, d):
        return Fraccion(n,d)

    def calcularResultado (self):
        return self.__numerador/self.__denominador

    def multiplicarPorNumero (self, num):
        return Fraccion(self.__numerador * num , self.__denominador)

    def multiplicarFracciones (self, fraccion2):
        return Fraccion(self.__numerador * fraccion2.getNumerador() , self.__denominador * fraccion2.getDenominador())

    def sumarFracciones (self, fraccion2):
        return Fraccion(self.__numerador * fraccion2.getDenominador() + self.__denominador * fraccion2.getNumerador(), self.__denominador * fraccion2.getDenominador())

    def restarFracciones (self, fraccion2) :
        return Fraccion(self.__numerador * fraccion2.getDenominador() - self.__denominador * fraccion2.getNumerador(), self.__denominador * fraccion2.getDenominador())

    def hallarMcd(self):
        dividendo = self.__numerador
        divisor = self.__denominador

        while dividendo % divisor != 0:
            resto = dividendo % divisor
            dividendo = divisor
            divisor = resto
        return int(divisor)

    def simplificarFraccion (self):
        mcd = fraccion.hallarMcd()
        self.__numerador /= mcd
        self.__denominador /= mcd

        return Fraccion(int(self.__numerador), int(self.__denominador))

    def __str__(self):
       return str(self.__numerador) + "/" + str(self.__denominador)

fraccion = Fraccion(2, 3)
fraccion2 = Fraccion(1, 5)

print("Fraccion:",fraccion)
print("La fraccion modificada es:",fraccion.modificar(1,2))
print("El resultado es: ", fraccion.calcularResultado())
print(fraccion,"* 2 =",fraccion.multiplicarPorNumero(2))
print(fraccion,"*" ,fraccion2, " =", fraccion.multiplicarFracciones(fraccion2))
print(fraccion,"+" ,fraccion2, " =", fraccion.sumarFracciones(fraccion2))
print(fraccion,"-" ,fraccion2, " =", fraccion.restarFracciones(fraccion2))
print("Fraccion simplificada",fraccion.simplificarFraccion())


